in/output management & mic recording

maxbolton opened the pull request:
- Change `voiceGroup` channel type from `VOICE` to `CHAT` in `AudioEngine`
- Replace `VOICE` with `CHAT` in `ChannelGroupType` enum in `AudioEngine.h`
- Add `AudioListenerComponent` class for handling microphone input and device management
- Update `AudioListenerComponent` constructor to take `PerspectiveCamera` reference
- Add methods to `AudioListenerComponent` for managing microphone recording and device lists
- Add `PlaySoundObj` method to `AudioSourceComponent`
- Update `CMakeLists.txt` to include `AudioListenerComponent.cpp`
- Add `GetForward` and `GetUp` methods to `Camera` class

Commit:
in/output management & mic recording

- Change `voiceGroup` channel type from `VOICE` to `CHAT` in `AudioEngine`
- Replace `VOICE` with `CHAT` in `ChannelGroupType` enum in `AudioEngine.h`
- Add `AudioListenerComponent` class for handling microphone input and device management
- Update `AudioListenerComponent` constructor to take `PerspectiveCamera` reference
- Add methods to `AudioListenerComponent` for managing microphone recording and device lists
- Add `PlaySoundObj` method to `AudioSourceComponent`
- Update `CMakeLists.txt` to include `AudioListenerComponent.cpp`
- Add `GetForward` and `GetUp` methods to `Camera` class

Alasdair-PB commented:
Could we remove the code from camera and keep it on the IComponent? IComponent has GetGameObject().GetTransform which you can use to get the forwards direction of the owning object and orientation. Then keep the Euler maths on this component? (I kinda just want to keep the camera code clean as I will remove player references from it entirely at some point)

Alasdair-PB commented:
If the camera is required for yaw let me know and I may try to rework camera into a component so it can accessed as a TryGetComponent so it uses the players transform/ rotation where it is a child gameobject of the player.

maxbolton commented:
Depends how we decide to make the game, if we go first person the forward and up vectors can just be taken from the object transform. This is just while the game is in third person. Can remove this from the branch and just keep it in my visual studio for testing for now.

Alasdair-PB commented:
Yeah if that's alright leaving it out for now may be best- thanks!

Commit:
Remove Quaternion dependency from Camera class

Removed #include "Quaternion.h" from Camera.h. Deleted GetForward and GetUp methods from Camera class, which used quaternion calculations for determining camera directions.

maxbolton commented:
fixed!

Commit:
Remove AudioObject.cpp from AudioEngine build

Unused .cpp file, all necessary methods are in the header

Commit:
Refactor AudioListenerComponent orientation methods

Introduce SetCamOrientation and SetPlayerOrientation methods in AudioListenerComponent to handle forward and up vector calculations. Update constructor and Update method to use these new methods.

Alasdair-PB reviewed (approved):
In the next PR change `LoadSound` to take a pointer created by the `SoundManager`, but looks good to me
